File: v2/java/part5/back_tracking/Boj2580.java
package part5.back_tracking;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Boj2580 {
    static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static final StringBuffer sb = new StringBuffer();
    static final List<int[]> zeros = new ArrayList<>();
    static final int N = 9;
    static StringTokenizer st;
    static int[][] map;

    public static void main(String[] args) throws IOException {
        map = new int[N][N];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                int token = Integer.parseInt(st.nextToken());
                if (token == 0) zeros.add(new int[] { i, j });
                map[i][j] = token;
            }
        }

        solve(0);
    }

    static void solve(int cur) {
        if (cur == zeros.size()) {
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++)
                    sb.append(map[i][j]).append(' ');
                sb.append('\n');
            }

            System.out.print(sb);
            System.exit(0);
        }

        int row = zeros.get(cur)[0];
        int col = zeros.get(cur)[1];

        for (int i = 1; i <= N; i++) {
            if (chk(row, col, i)) {
                map[row][col] = i;
                solve(cur + 1);
                map[row][col] = 0;
            }
        }
    }

    static boolean chk(int row, int col, int i) {
        for (int k = 0; k < N; k++)
            if (map[row][k] == i || map[k][col] == i)
                return false;

        int vx = (row / 3) * 3;
        int vy = (col / 3) * 3;

        for (int a = vx; a < vx + 3; a++)
            for (int b = vy; b < vy + 3; b++)
                if (map[a][b] == i)
                    return false;

        return true;
    }
}

File: v1/review/kotlin/src/main/kotlin/backtracking/boj_2580.kt
package backtracking

import java.util.StringTokenizer
import kotlin.system.exitProcess

/**
 * 스도쿠 https://www.acmicpc.net/problem/2580
 */

private const val N = 9
private val puzzle = Array(N) { IntArray(N) }
private val points = ArrayList<Pair<Int, Int>>()

private fun check(row: Int, col: Int, num: Int): Boolean {
    // 열 검사
    for (c in 0 until N)
        if (puzzle[row][c] == num)
            return false

    // 행 검사
    for (r in 0 until N)
        if (puzzle[r][col] == num)
            return false

    // 3 by 3 검사
    val tx = (row / 3) * 3
    val ty = (col / 3) * 3

    for (r in tx until tx + 3)
        for (c in ty until ty + 3)
            if (puzzle[r][c] == num)
                return false

    return true
}

private fun solve(n: Int) {
    if (n == points.size) {
        val sb = StringBuffer()

        for (i in 0 until N) {
            for (j in 0 until N)
                sb.append(puzzle[i][j]).append(' ')
            sb.append('\n')
        }

        print(sb)
        exitProcess(0)
    }

    for (i in 1..N) {
        val (x, y) = points[n]

        if (!check(x, y, i)) continue
        puzzle[x][y] = i
        solve(n + 1)
        puzzle[x][y] = 0
    }
}

fun boj_2580() = with(System.`in`.bufferedReader()) {
    for (i in 0 until N) {
        val st = StringTokenizer(readLine())

        for (j in 0 until N) {
            val p = st.nextToken().toInt()
            if (p == 0) points += Pair(i, j)
            puzzle[i][j] = p
        }
    }

    solve(0)
}